specs/sphinx-module-factory.md

```diff
@@ -0,0 +1,102 @@
+  1. User deploys on chain 1 using the strategy described in the previous section.
+  2. A malicious actor deploys a `SphinxModule` at the same `CREATE2` address on chain 2.
+  3. The user will not be able to deploy a Gnosis Safe at the same address on chain 2. It will revert because a `SphinxModule` already exists at the `CREATE2` address.
+- The address of a `SphinxModule` must be calculated via `CREATE2` using the following inputs:
+  - The address of the `SphinxModuleFactory`.
+  - The address of the Gnosis Safe contract that the `SphinxModule` belongs to.
+  - The address of the caller that deploys the `SphinxModule` through the `SphinxModuleFactory`.
+  - An arbitrary `uint256` nonce.
+- All of the behavior described in this specification must apply to [all Gnosis Safe contracts supported by Sphinx](TODO(end)).
+
+## Function-level Invariants
+
+#### `constructor`
+
+- Must deploy a `SphinxModule` implementation contract at a `CREATE2` address determined by the address of the `SphinxModuleFactory` and a `bytes32(0)` salt.
+- Must initialize the `SphinxModule` implementation so that nobody has permission to call its `approve` function.
+
+#### `function deploySphinxModule(address _safeProxy, uint256 _saltNonce) external returns (address sphinxModule);`
+
+- Must revert if a contract already exists at the `CREATE2` address.
+- A successful call must:
+  - Deploy an EIP-1167 proxy at the correct `CREATE2` address, using the `SphinxModule` implementation deployed in the `SphinxModuleFactory`'s constructor.
+  - Emit a `SphinxModuleDeployed` event in the `SphinxModuleFactory`.
+  - Initialize the `SphinxModule` using the correct Gnosis Safe address.
+  - Return the address of the `SphinxModule`.
+- A single caller must be able to deploy an arbitrary number of `SphinxModule`s by calling this function multiple times.
+
+#### `function deploySphinxModuleFromSafe(uint256 _saltNonce) external;`
+
+- Must revert if a contract already exists at the `CREATE2` address.
+- A successful call must:
+  - Deploy an EIP-1167 proxy at the correct `CREATE2` address, using the correct `SphinxModule` implementation deployed in the `SphinxModuleFactory`'s constructor.
+  - Emit a `SphinxModuleDeployed` event in the `SphinxModuleFactory`.
+  - Initialize the `SphinxModule` using the _caller's address_ as the Gnosis Safe address.
+- A single caller must be able to deploy an arbitrary number of `SphinxModule`s by calling this function multiple times.
+
+#### `function enableSphinxModuleFromSafe(uint256 _saltNonce) external;`
+
+- Must revert if not delegatecalled.
+- Must enable the correct `SphinxModule` as a module in the Gnosis Safe that triggered the delegatecall.
+- A single Gnosis Safe must be able to enable more than one `SphinxModule` by calling this function multiple times.
```

Comment:
It would be good to include a description of the successful call like you do for the above functions.